src/lightning/fabric/strategies/xla_fsdp.py

```diff
@@ -0,0 +1,403 @@
+            )
+
+    def remove_checkpoint(self, folderpath: _PATH) -> None:
+        """Remove checkpoint filepath from the filesystem.
+
+        Args:
+            filepath: Path to folder with full/sharded checkpoint(s)
+        """
+        rank = self.local_rank
+        world_size = self.world_size
+
+        # broadcast the path from rank 0 to ensure all the states are loaded from a common path
+        folderpath = self.broadcast(folderpath)
+        if not os.path.isdir(folderpath):
+            raise NotImplementedError(
+                "The `XLAFSDPStrategy` requires specifying the directory where to remove checkpoints."
+            )
+        if self._state_dict_type == "sharded":
+            self.checkpoint_io.remove_checkpoint(
+                os.path.join(folderpath, f"checkpoint_rank-{rank:08d}-of-{world_size:08d}.pth")
+            )
+        elif self._state_dict_type == "full":
+            self.checkpoint_io.remove_checkpoint(os.path.join(folderpath, "checkpoint_consolidated.pth"))
+        else:
+            raise ValueError(f"Unknown state_dict_type: {self._state_dict_type}")
+
+    def load_checkpoint(
+        self,
+        path: _PATH,
+        state: Optional[Union[Module, Optimizer, Dict[str, Union[Module, Optimizer, Any]]]] = None,
+        strict: bool = True,
+    ) -> Dict[str, Any]:
+        """Given a folder, load the contents from a checkpoint and restore the state of the given objects.
+
+        The strategy currently only supports saving and loading sharded checkpoints which are stored in form of a
+        directory of multiple files rather than a single file.
+        """
+        if not _TORCH_GREATER_EQUAL_2_0:
+            raise NotImplementedError(
+                "Saving and loading checkpoints with the `FSDPStrategy` is not supported in PyTorch < 2.0."
+                " Please upgrade `torch` or file an issue: `https://github.com/Lightning-AI/lightning/issues`."
+            )
+        if not state:
+            raise ValueError(
+                f"Got XLAFSDPStrategy.load_checkpoint(..., state={state!r}) but a state with at least "
+                f" a model instance to reload is required. Pass it in like so:"
+                " FSDPStrategy.load_checkpoint(..., state={'model': model, ...})"
+            )
+
+        # broadcast the path from rank 0 to ensure all the states are loaded from a common path
+        path = self.broadcast(path)
+        if not os.path.isdir(path):
+            raise NotImplementedError(
+                f"The path `{path}` is a file or does not exist, but the `XLAFSDPStrategy` currently only supports loading from a"
+                " checkpoint(s) in a directory."
+            )
+        rank = self.local_rank
+        world_size = self.world_size
+
+        from torch_xla.distributed.fsdp import XlaFullyShardedDataParallel as XLAFSDP
+
+        modules = {key: module for key, module in state.items() if isinstance(module, XLAFSDP)}
```

Comment:
In regular FSDP, it's supported that the root is not FSDP-wrapped. So the logic to find valid modules checks whether any submodule is FSDP-wrapped: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/strategies/fsdp.py#L524

Should we do the same here?

Reply:
afaik the entire model will have to be wrapped so we do not need to add logic when only submodules are wrapped. cc @Liyang90 and @alexwertheim please correct me if I'm wrong

Reply:
At https://github.com/Lightning-AI/lightning/pull/18126/files#diff-3908a573abf00ae5f37061f214f2a3c2616b6591e0c96206b9f48b4c7ab49ea4R78 you added logic to avoid wrapping the root if a submodule is already wrapped. You also added a test for that: https://github.com/Lightning-AI/lightning/pull/18126/files#diff-719d0c95d7fc3c4cb9876bb0f5ea512a57889d8f728a4d0973eb196b167f42b6R220

This is in line with what was done to regular FSDP: https://github.com/Lightning-AI/lightning/pull/18054.

So we should clarify if XLA FSDP supports this sort of behaviour

Reply:
Sorry for the confusion. yes, FSDP in Pytorch XLA does support wrapping submodules with XLAFSDP. However, the outer model will have to be wrapped too, which makes the logic in the current code wrong. I will create a fix for this

Reply:
Instead of the current logic, what I will change is that the auto_wrap_policy will get ignored so that we just wrap the outside of the model

Reply:
New logic:
```
def setup_module(self, module: Module) -> Module:
        from torch_xla.distributed.fsdp import XlaFullyShardedDataParallel as XLAFSDP

        if any(isinstance(mod, XLAFSDP) for mod in module.modules()) and "auto_wrap_policy" in self._fsdp_kwargs:
            rank_zero_warn(
                "A XLAFSDP `auto_wrap_policy` is set, but at least one submodule is already wrapped."
                " The policy will be ignored."
            )
            del self._fsdp_kwargs["auto_wrap_policy"]

        if self._sync_module_states:
            if _XLA_GREATER_EQUAL_2_1:
                from torch_xla.core.xla_model import broadcast_master_param
            else:
                from torch_xla.experimental.pjrt import broadcast_master_param

            broadcast_master_param(module)

        module = XLAFSDP(
            module=module,
            **self._fsdp_kwargs,
        )

        return module
```

cc @carmocca, feel free to close this comment if this works

Reply:
Won't wrapping the root forcefully shard any parameters that are in the root? Wouldn't that be undesirable in the case where the user wants to specifically select which layers to wrap but the root contains parameters? Do you have a link that mentions this requirement?

Reply:
Maybe the source code here will add clarity: https://github.com/pytorch/xla/blob/master/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py#L205-L209